distsql: add infrastructure for running simple flows without goroutines

RaduBerinde opened the pull request:
We can run certain simple local flows in "serial" mode, where each processor
dumps its output into a row buffer. This will be used for very small queries
(the kind where we know exactly how many rows we are going to read).

Introducing a rowSource interface and changing joinReader to use that as input.

Benchmarks (after is with the test flag enabled):

```
Scan1_Cockroach-4                    140µs ± 5%    162µs ±23%    +15.64%    (p=0.002 n=9+9)
Scan10_Cockroach-4                   169µs ± 2%    191µs ±28%    +13.07%   (p=0.004 n=8+10)
Scan100_Cockroach-4                  390µs ± 9%    438µs ±15%    +12.36%    (p=0.000 n=9+9)
Scan1000_Cockroach-4                2.42ms ± 7%   2.60ms ± 7%     +7.34%   (p=0.016 n=10+8)
Scan10000_Cockroach-4               21.8ms ± 3%   24.4ms ± 4%    +11.81%  (p=0.000 n=10+10)
Scan1000Limit1_Cockroach-4           154µs ±12%    154µs ± 6%       ~     (p=0.971 n=10+10)
Scan1000Limit10_Cockroach-4          171µs ± 4%    177µs ± 1%     +3.84%   (p=0.001 n=10+8)
Scan1000Limit100_Cockroach-4         389µs ± 7%    417µs ± 8%     +7.25%   (p=0.008 n=9+10)
```

(I'm not including the FilterLimit benchmarks - those are not the kind of queries that
will be run in this mode, as we don't know how many rows we need to read).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7300)

<!-- Reviewable:end -->

tschottdorf commented:
Any word on the change in allocs for your benchmark numbers above?

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/distsql/base.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKifrwq44BI3-s_mNd2:-KKifrwq44BI3-s_mNd3:824276694) ([raw file](https://github.com/cockroachdb/cockroach/blob/066d8393424fd680f85e27d3ff843d2ab0d4dae1/sql/distsql/base.go#L130)):*
> ```Go
> // RowBuffer is an implementation of rowReceiver that buffers (accumulates)
> // results in memory, as well as an implementation of rowSender that returns
> // rows from a row buffer.
> ```

Isn't this type a little problematic (or at least unintuitive)? If you have one goroutine calling `PushRow` and the other "end" calls `ReceiveRow`, the receiver and the pusher race which could lead to missing out on rows. Shouldn't `NextRow()` block until `tr.closed` is true (which of course isn't that easy)? Or the semantics/intended usage need to be documented somewhat more clearly - the interfaces don't make these subtleties clear. I know that the current usage is unproblematic (as you're completely running the flow before collecting the results), just something to keep in mind as the codebase grows.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-:-KKifCiYCHhGDkuZPGUT:1587904877)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/distsql/base.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKifrwq44BI3-s_mNd2:-KKihbpUZEjG78nXxrmE:-463728912) ([raw file](https://github.com/cockroachdb/cockroach/blob/066d8393424fd680f85e27d3ff843d2ab0d4dae1/sql/distsql/base.go#L130)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Isn't this type a little problematic (or at least unintuitive)? If you have one goroutine calling `PushRow` and the other "end" calls `ReceiveRow`, the receiver and the pusher race which could lead to missing out on rows. Shouldn't `NextRow()` block until `tr.closed` is true (which of course isn't that easy)? Or the semantics/intended usage need to be documented somewhat more clearly - the interfaces don't make these subtleties clear. I know that the current usage is unproblematic (as you're completely running the flow before collecting the results), just something to keep in mind as the codebase grows.
</details>
The point is not to use it as both a receiver and a sender at the same time - when using multiple goroutines, RowChannel is used. This is when you want to run the processors serially - first run a processor that puts all the rows in a RowBuffer, then start and run a processer that consumes them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[sql/distsql/base.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKifrwq44BI3-s_mNd2:-KKii3IbTzaWHiFrdgGL:261417392) ([raw file](https://github.com/cockroachdb/cockroach/blob/066d8393424fd680f85e27d3ff843d2ab0d4dae1/sql/distsql/base.go#L130)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> The point is not to use it as both a receiver and a sender at the same time - when using multiple goroutines, RowChannel is used. This is when you want to run the processors serially - first run a processor that puts all the rows in a RowBuffer, then start and run a processer that consumes them.
</details>
Agreed this is not clearly spelled out though, will improve the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/dist_sql_node.go, line 45 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKiYUC1aolyzibH4CMM:-KKiYUC1aolyzibH4CMN:171676767) ([raw file](https://github.com/cockroachdb/cockroach/blob/066d8393424fd680f85e27d3ff843d2ab0d4dae1/sql/dist_sql_node.go#L45)):*
> ```Go
> 
> 	// rowBuf accumulates the results of the flow.
> 	rowBuf distsql.RowBuffer
> ```

explain that this is not always used... Or is it?
Or did you mean `RowSource` here?


Also, mind explaining how this member is related to the `flow` above ? Should't it be a member of the `Flow` interface?

---

*[sql/dist_sql_node.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKiXozX9Huev-gqk5-z:-KKiXozYNjDvVCX9Xw7T:-2131577103) ([raw file](https://github.com/cockroachdb/cockroach/blob/066d8393424fd680f85e27d3ff843d2ab0d4dae1/sql/dist_sql_node.go#L87)):*
> ```Go
> func (n *distSQLNode) Next() (bool, error) {
> 	if !n.ranFlow {
> 		n.flow.RunSync()
> ```

ha? doesn't sync/non-sync need to be configured somehow?

---

*[sql/distsql/base.go, line 42 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKiYGfnrS2GNFj7Su5z:-KKiYGfnrS2GNFj7Su6-:911403665) ([raw file](https://github.com/cockroachdb/cockroach/blob/066d8393424fd680f85e27d3ff843d2ab0d4dae1/sql/distsql/base.go#L42)):*
> ```Go
> }
> 
> type rowSource interface {
> ```

comment pls

---

*[sql/distsql/base.go, line 147 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKi_4Rm9POEXxVPwaeE:-KKi_4Rn1xUXvSqwUcVO:171911058) ([raw file](https://github.com/cockroachdb/cockroach/blob/066d8393424fd680f85e27d3ff843d2ab0d4dae1/sql/distsql/base.go#L147)):*
> ```Go
> 
> // PushRow is part of the rowReceiver interface.
> func (tr *RowBuffer) PushRow(row sqlbase.EncDatumRow) bool {
> ```

what's `tr`?

---

*[sql/distsql/tablereader_test.go, line 33 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKi_KAC9yyo0z-55Rxc:-KKi_KAC9yyo0z-55Rxd:677880290) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f4721d7a66ca3e06cd2e65429597491d3ae192a/sql/distsql/tablereader_test.go#L33)):*
> ```Go
> )
> 
> // testingReceiver is an implementation of rowReceiver that accumulates
> ```

nice

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[sql/dist_sql_node.go, line 87 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKiXozX9Huev-gqk5-z:-KKimI7i_mrl-g621s7A:-998525365) ([raw file](https://github.com/cockroachdb/cockroach/blob/066d8393424fd680f85e27d3ff843d2ab0d4dae1/sql/dist_sql_node.go#L87)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> ha? doesn't sync/non-sync need to be configured somehow?
</details>
Ah, yeah my initial intention was to switch over to this but we will need both paths. Will fix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Updated. We now support both sync and async modes.

```
name \ time/op                      /tmp/bench-0  /tmp/bench-1   /tmp/bench-2
Scan1_Cockroach-4                     149µs ± 9%     154µs ±11%    283µs ± 5%
Scan10_Cockroach-4                    201µs ±57%     187µs ± 7%    314µs ± 5%
Scan100_Cockroach-4                   420µs ±14%     429µs ± 9%    585µs ±11%
Scan1000_Cockroach-4                 2.43ms ± 4%    2.71ms ±12%   2.68ms ± 7%
Scan10000_Cockroach-4                22.4ms ± 4%    25.6ms ±12%   23.3ms ±10%
Scan1000Limit1_Cockroach-4            147µs ± 5%     156µs ± 3%    296µs ± 3%
Scan1000Limit10_Cockroach-4           176µs ± 5%     192µs ± 1%    324µs ± 3%
Scan1000Limit100_Cockroach-4          397µs ± 6%     437µs ± 3%    574µs ± 6%
Scan10000FilterLimit1_Cockroach-4     303µs ± 7%   10330µs ± 5%   1284µs ±12%
Scan10000FilterLimit10_Cockroach-4    346µs ± 2%   10312µs ± 4%   1338µs ±13%
Scan10000FilterLimit50_Cockroach-4   2.31ms ± 2%   10.25ms ± 5%   2.29ms ± 2%

name \ alloc/op                     /tmp/bench-0  /tmp/bench-1   /tmp/bench-2
Scan1_Cockroach-4                    15.5kB ± 0%    20.2kB ± 0%   20.8kB ± 0%
Scan10_Cockroach-4                   19.0kB ± 0%    24.8kB ± 0%   24.3kB ± 0%
Scan100_Cockroach-4                  49.8kB ± 0%    69.9kB ± 0%   59.8kB ± 0%
Scan1000_Cockroach-4                  317kB ± 0%     466kB ± 0%    370kB ± 0%
Scan10000_Cockroach-4                5.53MB ± 0%    7.57MB ± 0%   6.01MB ± 0%
Scan1000Limit1_Cockroach-4           16.0kB ± 0%    20.8kB ± 0%   21.5kB ± 0%
Scan1000Limit10_Cockroach-4          19.4kB ± 0%    25.4kB ± 0%   24.9kB ± 0%
Scan1000Limit100_Cockroach-4         50.1kB ± 0%    70.4kB ± 0%   60.2kB ± 0%
Scan10000FilterLimit1_Cockroach-4    47.9kB ± 0%  2357.0kB ± 0%  413.7kB ± 0%
Scan10000FilterLimit10_Cockroach-4   54.2kB ± 0%  2361.7kB ± 0%  420.5kB ± 0%
Scan10000FilterLimit50_Cockroach-4    461kB ± 0%    2215kB ± 0%    481kB ± 0%

name \ allocs/op                    /tmp/bench-0  /tmp/bench-1   /tmp/bench-2
Scan1_Cockroach-4                       194 ± 0%       222 ± 0%      222 ± 0%
Scan10_Cockroach-4                      221 ± 0%       262 ± 0%      250 ± 0%
Scan100_Cockroach-4                     417 ± 0%       557 ± 0%      452 ± 0%
Scan1000_Cockroach-4                  2.30k ± 0%     3.40k ± 0%    2.39k ± 0%
Scan10000_Cockroach-4                 21.1k ± 0%     31.8k ± 0%    21.8k ± 0%
Scan1000Limit1_Cockroach-4              201 ± 0%       233 ± 0%      233 ± 0%
Scan1000Limit10_Cockroach-4             227 ± 0%       272 ± 0%      259 ± 0%
Scan1000Limit100_Cockroach-4            423 ± 0%       567 ± 0%      462 ± 0%
Scan10000FilterLimit1_Cockroach-4       501 ± 0%      8646 ± 0%     1251 ± 0%
Scan10000FilterLimit10_Cockroach-4      555 ± 0%      8690 ± 0%     1306 ± 0%
Scan10000FilterLimit50_Cockroach-4    1.79k ± 0%     8.85k ± 0%    1.94k ± 0%
```

---

Review status: 0 of 11 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-:-KKjCt-Rr0I1A9ezVdDS:1980178764)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
distsql: add infrastructure for running simple flows without goroutines

We can run certain simple local flows in "serial" mode, where each processor
dumps its output into a row buffer. This will be used for very small queries
(the kind where we know exactly how many rows we are going to read).

Introducing a rowSource interface and changing joinReader to use that as input.

Benchmarks 0 is off (default), 1 is sync mode, 2 is async mode.

```
name \ time/op                      /tmp/bench-0  /tmp/bench-1   /tmp/bench-2
Scan1_Cockroach-4                     149µs ± 9%     154µs ±11%    283µs ± 5%
Scan10_Cockroach-4                    201µs ±57%     187µs ± 7%    314µs ± 5%
Scan100_Cockroach-4                   420µs ±14%     429µs ± 9%    585µs ±11%
Scan1000_Cockroach-4                 2.43ms ± 4%    2.71ms ±12%   2.68ms ± 7%
Scan10000_Cockroach-4                22.4ms ± 4%    25.6ms ±12%   23.3ms ±10%
Scan1000Limit1_Cockroach-4            147µs ± 5%     156µs ± 3%    296µs ± 3%
Scan1000Limit10_Cockroach-4           176µs ± 5%     192µs ± 1%    324µs ± 3%
Scan1000Limit100_Cockroach-4          397µs ± 6%     437µs ± 3%    574µs ± 6%
Scan10000FilterLimit1_Cockroach-4     303µs ± 7%   10330µs ± 5%   1284µs ±12%
Scan10000FilterLimit10_Cockroach-4    346µs ± 2%   10312µs ± 4%   1338µs ±13%
Scan10000FilterLimit50_Cockroach-4   2.31ms ± 2%   10.25ms ± 5%   2.29ms ± 2%

name \ alloc/op                     /tmp/bench-0  /tmp/bench-1   /tmp/bench-2
Scan1_Cockroach-4                    15.5kB ± 0%    20.2kB ± 0%   20.8kB ± 0%
Scan10_Cockroach-4                   19.0kB ± 0%    24.8kB ± 0%   24.3kB ± 0%
Scan100_Cockroach-4                  49.8kB ± 0%    69.9kB ± 0%   59.8kB ± 0%
Scan1000_Cockroach-4                  317kB ± 0%     466kB ± 0%    370kB ± 0%
Scan10000_Cockroach-4                5.53MB ± 0%    7.57MB ± 0%   6.01MB ± 0%
Scan1000Limit1_Cockroach-4           16.0kB ± 0%    20.8kB ± 0%   21.5kB ± 0%
Scan1000Limit10_Cockroach-4          19.4kB ± 0%    25.4kB ± 0%   24.9kB ± 0%
Scan1000Limit100_Cockroach-4         50.1kB ± 0%    70.4kB ± 0%   60.2kB ± 0%
Scan10000FilterLimit1_Cockroach-4    47.9kB ± 0%  2357.0kB ± 0%  413.7kB ± 0%
Scan10000FilterLimit10_Cockroach-4   54.2kB ± 0%  2361.7kB ± 0%  420.5kB ± 0%
Scan10000FilterLimit50_Cockroach-4    461kB ± 0%    2215kB ± 0%    481kB ± 0%

name \ allocs/op                    /tmp/bench-0  /tmp/bench-1   /tmp/bench-2
Scan1_Cockroach-4                       194 ± 0%       222 ± 0%      222 ± 0%
Scan10_Cockroach-4                      221 ± 0%       262 ± 0%      250 ± 0%
Scan100_Cockroach-4                     417 ± 0%       557 ± 0%      452 ± 0%
Scan1000_Cockroach-4                  2.30k ± 0%     3.40k ± 0%    2.39k ± 0%
Scan10000_Cockroach-4                 21.1k ± 0%     31.8k ± 0%    21.8k ± 0%
Scan1000Limit1_Cockroach-4              201 ± 0%       233 ± 0%      233 ± 0%
Scan1000Limit10_Cockroach-4             227 ± 0%       272 ± 0%      259 ± 0%
Scan1000Limit100_Cockroach-4            423 ± 0%       567 ± 0%      462 ± 0%
Scan10000FilterLimit1_Cockroach-4       501 ± 0%      8646 ± 0%     1251 ± 0%
Scan10000FilterLimit10_Cockroach-4      555 ± 0%      8690 ± 0%     1306 ± 0%
Scan10000FilterLimit50_Cockroach-4    1.79k ± 0%     8.85k ± 0%    1.94k ± 0%
``

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 11 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[sql/dist_sql_node.go, line 36 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKkG7yTK48d0qwhf1Vl:-KKkG7yTK48d0qwhf1Vm:-1182412627) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dfea5f09953906351184aad5ea7024771bec089/sql/dist_sql_node.go#L36)):*
> ```Go
> 	ordering orderingInfo
> 
> 	// syncMode indicates the mode in which we run the associated flow:
> ```

comment sounds like it came from another version

---

*[sql/dist_sql_node.go, line 201 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-KKkHHJFJjEMN5R4m1PL:-KKkHHJG1-zOAY1q-jhV:763143240) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dfea5f09953906351184aad5ea7024771bec089/sql/dist_sql_node.go#L201)):*
> ```Go
> // scanNodeToDistSQL creates a flow and distSQLNode that correspond to a
> // scanNode.
> func scanNodeToDistSQL(n *scanNode, syncMode bool) (*distSQLNode, error) {
> ```

syncMode deserves a comment. Or, make it an enum.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7300#-:-KKkHXqxnbP4KK37dPV9:1417088721)*
<!-- Sent from Reviewable.io -->
